helix-term/src/ui/picker.rs

```diff
@@ -30,16 +34,21 @@ pub struct Picker<T> {
     cursor: usize,
     // pattern: String,
     prompt: Prompt,
+    // Caches paths to docs to line number to view
+    preview_cache: HashMap<PathBuf, (Document, HashMap<usize, View>)>,
```

Comment:
This also presumes that a picker always picks files and ends up with a Document. At the very least the key here should probably be `T`.

Reply:
It might also make more sense to make a new `struct FilePicker` that contains two components, a Picker and a Preview. That way the base picker is less complex.

Reply:
`T` may not implement `Hash` which is the case for goto, symbol, etc. Using `FilePicker` might solve that problem, I'll give it a go.

Reply:
Yeah I think composing is a good way to go, similar to how the picker itself is technically a prompt with a list.

Also, what do you think about a layout like this?

```
 ---------   ---------
|         | |         |
|---------| |         |
|         | |         |
|         | |         |
|         | |         |
 ---------   ---------
```

The FilePicker would then be a picker on the left with 50% of the width, and a preview on the right.

Reply:
Changed the layout and added a specialized `FilePicker`. A generic previewed picker which can render any component would be better, but it's hard to implement as of now.